### PR TITLE
Added option to impute missing values during training

### LINKED ISSUE #301: Impute missing values after the train/test split
To prevent potential for data leakage (see <https://arxiv.org/pdf/2207.07048.pdf>).

{mice} package may be useful.

### PR DESCRIPTION
**Pull request by megancoden**
## Issues
- Resolves #301  .

## Change(s) made
- in preprocess.R, we added an option called impute_in_preprocessing that defaults to True. When it is true, it calls the impute function, which performs the imputation on the data. When it is set to false, this step gets skipped.
- We created a file called impute.R where we moved code from preprocess.R that would be shared between files. We created two functions: impute(), which takes in a dataset and performs the imputation, and prep_data(), which gets the dataset into the proper format to perform imputation.
- In run_ml, we added an option called impute_in_training, which defaults to False. When it is true, it calls prep_data() and performs the imputation step following the train/test data split. When it is false, it does not modify the input dataset.
- We added test cases to test-run_ml.R and in test-preprocess.R to ensure that our functionality worked as expected.

## Checklist

(~Strikethrough~ any points that are not applicable.)

- [X] Write unit tests for any new functionality or bug fixes.
~~- [ ] Update docs if there are any API changes:~~
  ~~- [ ] roxygen comments~~
  ~~- [ ] vignettes~~
- [X] Update `NEWS.md` if this includes any user-facing changes. 
- [X] The check workflow succeeds on your most recent commit. **This is always required before the PR can be merged.**

### CODE REVIEW

File: R/run_ml.R [hunk 1]
```diff
@@ -144,6 +146,7 @@ run_ml <-
            group_partitions = NULL,
            corr_thresh = 1,
            seed = NA,
+           impute_in_training = FALSE,
```
Comment (kelly-sovacool) [inline R/run_ml.R:149]: I would name this parameter `impute_after_split` to make it clear that the imputation occurs after the train/test split and before model training and testing.

File: R/preprocess.R [hunk 2]
```diff
@@ -70,34 +73,15 @@ preprocess_data <- function(dataset, outcome_colname,
...
-
-  nv_feats <- process_novar_feats(features, progbar = progbar)
-  pbtick(progbar)
-  split_feats <- process_cat_feats(nv_feats$var_feats, progbar = progbar)
-  pbtick(progbar)
-  cont_feats <- process_cont_feats(split_feats$cont_feats, method)
-  pbtick(progbar)
-
+  
+  processed_data <- prep_data(dataset, outcome_colname, prefilter_threshold, method, impute_in_preprocessing, to_numeric)
```
Comment (kelly-sovacool) [inline R/preprocess.R:77]: Why do you think moving this code to a separate `prep_data` function is necessary?

File: R/run_ml.R [hunk 3]
```diff
@@ -201,30 +204,43 @@ run_ml <-
...
+      train_processed_feats <- train_processed_data$processed_feats
+      split_dat <- train_processed_data$split_dat
+      train_data <- dplyr::bind_cols(split_dat$outcome, train_processed_feats) %>%
+        dplyr::as_tibble()
+      test_processed_data <- prep_data(test_data, outcome_colname, prefilter_threshold=1, method=c("center", "scale"), impute_in_preprocessing=TRUE, to_numeric=TRUE)
+      test_processed_feats <- test_processed_data$processed_feats
+      split_dat <- test_processed_data$split_dat
+      test_data <- dplyr::bind_cols(split_dat$outcome, test_processed_feats) %>%
+        dplyr::as_tibble()
+    }
```
Comment (kelly-sovacool) [inline R/run_ml.R:222]: As you now have it, `prep_data()` does more than just impute missing values -- it also does other preprocessing steps as advertised by `preprocess_data()`. 

So it seems like there are two potential implementations of this feature:

A) Users can opt to do all preprocessing, including imputation, after the train/test split.
B) Users can opt to do only imputation and not other preprocessing after the train/test split. In this instance, presumably they perform all preprocessing minus imputation before calling `run_ml`.
C) As it is now, users could perform the train/test split on their own, do preprocessing on the splits separately (which includes imputation), and assign the train/test sets in `run_ml()` explicitly by providing a vector of indices to `train_frac`.

This opens a wider discussion about how we want the API to work, and how much flexibility we want to give users. 

What this code currently provides is an easy way to do option A. What I initially had in mind for this PR is option B, because option A could be done via option C.

File: tests/testthat/test-run_ml.R [hunk 4]
```diff
@@ -1,3 +1,4 @@
+# TODO: add tests for if impution param is set to true
```
Comment (kelly-sovacool) [inline tests/testthat/test-run_ml.R:1]: remove this TODO when you've completed it

File: R/preprocess.R (hunk 2 above)
Comment (kelly-sovacool) [reply R/preprocess.R:77]: Oh, I see that you also called it in `run_ml` below. I think we don't need `prep_data', because in `run_ml` it should be possible to _only_ impute missing values after the split and _not_ do any other preprocessing steps there. (see other comment below)

**Comment by kelly-sovacool** [comment]
These are API changes, so we'll need to document how to use them in the 'preprocess' vignette and roxygen comments.

File: R/impute.R [hunk 5]
```diff
@@ -0,0 +1,49 @@
+impute <- function(transformed_cont, n_missing) {
```
Comment (kelly-sovacool) [inline R/impute.R:1]: This function needs a roxygen docstring -- see other functions for examples.

https://r-pkgs.org/man.html#roxygen2-basics

**Comment by kelly-sovacool** [comment]
Tip: You can run `devtools::check()` to see if your local version of the package passes the checks successfully.

https://r-pkgs.org/whole-game.html#check

File: R/preprocess.R (hunk 2 above)
Comment (megancoden) [reply R/preprocess.R:77]: I see what you are saying. We removed the prep_data function and instead added the step to compute n_missing to the impute function.

File: R/run_ml.R (hunk 3 above)
Comment (megancoden) [reply R/run_ml.R:222]: We reworked the code so it does what you described in option B now.

**Comment by megancoden** [comment]
> These are API changes, so we'll need to document how to use them in the 'preprocess' vignette and roxygen comments.

Would you like us to include information about how to use it in the run_ml function in the 'preprocess' vignette, or should this be included elsewhere?

**Comment by kelly-sovacool** [comment]
> > These are API changes, so we'll need to document how to use them in the 'preprocess' vignette and roxygen comments.
> 
> Would you like us to include information about how to use it in the run_ml function in the 'preprocess' vignette, or should this be included elsewhere?

I would discuss it in a new subheading the preprocess vignette, probably nested under the heading "Replace missing continuous data with the median value of that feature" (also maybe rephrase this heading to make it more succinct). Show how to call both preprocess_data and run_ml in one code chunk.
